pkg/kubectl/resource/categories.go

```diff
@@ -36,6 +36,58 @@ func (e SimpleCategoryExpander) Expand(category string) ([]schema.GroupResource,
 	return ret, ok
 }
 
+type discoveryCategoryExpander struct {
+	fallbackExpander CategoryExpander
+	discoveryClient  discovery.DiscoveryInterface
+}
+
+// NewDiscoveryCategoryExpander returns a category expander that makes use of the "categories" fields from
+// the API, found through the discovery client. In case of any error or no category found (which likely
+// means we're at a cluster prior to categories support, fallback to the expander provided.
+func NewDiscoveryCategoryExpander(fallbackExpander CategoryExpander, client discovery.DiscoveryInterface) (discoveryCategoryExpander, error) {
+	if client == nil {
+		return discoveryCategoryExpander{}, errors.New("Please provide discovery client to shortcut expander")
+	}
+	return discoveryCategoryExpander{fallbackExpander: fallbackExpander, discoveryClient: client}, nil
+}
+
+func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResource, bool) {
+	apiResourceLists, err := e.discoveryClient.ServerResources()
+	if err != nil {
+		return e.fallbackExpander.Expand(category)
+	}
+
+	discoveredExpansions := map[string][]schema.GroupResource{}
+
+	for _, apiResourceList := range apiResourceLists {
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
```

Comment:
this particular error is bad, I would return an error instead of falling back.

Reply:
@deads2k should we change the `CategoryExpander` interface signature? No error returned in `Expand` ATM.

Reply:
> @deads2k should we change the CategoryExpander interface signature? No error returned in Expand ATM.

hmmm... ok.  I can deal with a fallback for 1.8 to avoid too many ripples at this stage.